### PR TITLE
[Merged by Bors] - feat(GroupTheory/GroupAction/Basic): various orbit / quotient lemmas

### PR DESCRIPTION
**Pull request by jsm28**
Add various lemmas about `orbitRel.Quotient`, mostly concerning `orbitRel.Quotient.orbit`, many concerning orbits for the action of a subgroup.  Along with the lemmas there are instances

```lean
instance (x : orbitRel.Quotient G α) : MulAction G x.orbit where
```

and

```lean
instance (x : orbitRel.Quotient G α) : IsPretransitive G x.orbit where
```

both of which already have analogues for `MulAction.orbit` as opposed to `orbitRel.Quotient.orbit` - and lemmas about the above action on an orbit.

From AperiodicMonotilesLean.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/GroupTheory/GroupAction/Basic.lean [hunk 1]
```diff
@@ -533,6 +533,102 @@ theorem orbitRel.Quotient.orbit_eq_orbit_out (x : orbitRel.Quotient G α)
...
+@[to_additive]
+lemma orbitRel.Quotient.mem_subgroup_orbit_iff {H : Subgroup G} {x : orbitRel.Quotient G α}
+    {a b : x.orbit} : a ∈ MulAction.orbit H b ↔ (a : α) ∈ MulAction.orbit H (b : α) := by
+  refine ⟨fun h ↦ ?_, fun h ↦ ?_⟩
+  · rcases h with ⟨g, rfl⟩
+    simp_rw [Submonoid.smul_def, Subgroup.coe_toSubmonoid, orbit.coe_smul, ← Submonoid.smul_def]
+    exact MulAction.mem_orbit _ g
+  · rcases h with ⟨g, h⟩
+    simp_rw [Submonoid.smul_def, Subgroup.coe_toSubmonoid, ← orbit.coe_smul,
+             ← Submonoid.smul_def, ← Subtype.ext_iff] at h
```
Comment (j-loreaux) [inline Mathlib/GroupTheory/GroupAction/Basic.lean:604]:
```suggestion
    simp_rw [Submonoid.smul_def, Subgroup.coe_toSubmonoid, ← orbit.coe_smul,
      ← Submonoid.smul_def, ← Subtype.ext_iff] at h
```

**Review by j-loreaux — COMMENTED** [review]
Probably you should just open the `orbitRel` namespace. (or even `orbitRel.Quotient`.

File: Mathlib/GroupTheory/GroupAction/Basic.lean [hunk 2]
```diff
@@ -533,6 +533,102 @@ theorem orbitRel.Quotient.orbit_eq_orbit_out (x : orbitRel.Quotient G α)
...
+  simp_rw [mem_orbit_symm (a₂ := c)]
+  convert Iff.rfl using 2
+  rw [orbit_eq_iff]
+  suffices hb : ↑b ∈ orbitRel.Quotient.orbit (⟦a⟧ : orbitRel.Quotient H x.orbit) by
+    rw [orbitRel.Quotient.orbit_eq_orbit_out (⟦a⟧ : orbitRel.Quotient H x.orbit) Quotient.out_eq']
+       at hb
+    rw [← orbitRel.Quotient.mem_subgroup_orbit_iff]
+    convert hb using 1
+    rw [orbit_eq_iff]
+    change @Setoid.r _ (orbitRel H _) _ _
```
[the remark this answers is not in the document]
Comment (jsm28) [reply Mathlib/GroupTheory/GroupAction/Basic.lean]: I've added `orbitRel_r_apply`.

In general I may not know the best way to make use of the API for quotients; I seem to find a lot that I have things in one form (`Setoid.r`, `Setoid.rel`, something defeq to those for the particular setoid, `≈`, ..., or `mk''` versus `mk` for quotients), and need to keep changing to a different form because the particular lemma I want to use next is only available for some of the forms, and also often find that key information is implicit and not shown in the infoview (making it hard to follow the infoview state, which e.g. just says `Setoid.r` with no information about which setoid is involved, as well as often requiring use of `@`), especially when, as here, dealing with multiple quotients for the same type (orbits for a group and a subgroup thereof, in this case).

**Comment by jsm28** [comment]
> Probably you should just open the `orbitRel` namespace. (or even `orbitRel.Quotient`.

There's enough overlap with the `MulAction` namespace (that we're inside here) that I suspect having both open would be confusing.

File: Mathlib/GroupTheory/GroupAction/Basic.lean [hunk 3]
```diff
@@ -411,6 +411,10 @@ theorem orbitRel_apply {a b : α} : (orbitRel G α).Rel a b ↔ a ∈ orbit G b
 #align mul_action.orbit_rel_apply MulAction.orbitRel_apply
 #align add_action.orbit_rel_apply AddAction.orbitRel_apply
 
+@[to_additive]
+lemma orbitRel_r_apply {a b : α} : @Setoid.r _ (orbitRel G _) a b ↔ a ∈ orbit G b :=
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/GroupAction/Basic.lean:415]: I believe the following works
```suggestion
lemma orbitRel_r_apply {a b : α} : (orbitRel G _).r a b ↔ a ∈ orbit G b :=
```
since Lean 4 tries to fill dot notation arguments to any kind of parameters, not just explicit

File: Mathlib/GroupTheory/GroupAction/Basic.lean [hunk 4]
```diff
@@ -533,6 +537,100 @@ theorem orbitRel.Quotient.orbit_eq_orbit_out (x : orbitRel.Quotient G α)
...
+    rw [orbitRel.Quotient.mem_orbit, Quotient.eq''] at yh zh
+    rcases yh with ⟨g, rfl⟩
+    rcases zh with ⟨h, rfl⟩
+    refine ⟨h * g⁻¹, ?_⟩
+    ext
+    simp [mul_smul]
+
+@[to_additive]
+lemma orbitRel.Quotient.mem_subgroup_orbit_iff {H : Subgroup G} {x : orbitRel.Quotient G α}
+    {a b : x.orbit} : a ∈ MulAction.orbit H b ↔ (a : α) ∈ MulAction.orbit H (b : α) := by
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/GroupAction/Basic.lean:601]: Maybe turn this one around and `@[simp, norm_cast]` it?

**Review by YaelDillies — COMMENTED** [review]
I feel your pain regarding quotients and setoids 😢

File: Mathlib/GroupTheory/GroupAction/Basic.lean (hunk 4 above)
Comment (jsm28) [reply Mathlib/GroupTheory/GroupAction/Basic.lean]: Done.

**Review by YaelDillies — APPROVED** [review]
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge
